packages/typed-ipc/src/main.ts

```diff
@@ -8,42 +8,62 @@ export * from './types'
  * Typed listener for Electron `ipcMain`.
  */
 export class IpcListener<T extends IpcEventMap> {
-  private listeners: string[] = []
-  private handlers: string[] = []
+  private listeners: Map<string, Set<(...args: any[]) => void>> = new Map()
+  private handlers: Set<string> = new Set()
 
   /**
    * Listen to `channel`.
+   * @returns A function to remove the listener.
    */
   on<E extends keyof ExtractArgs<T>>(
     channel: Extract<E, string>,
     listener: (e: Electron.IpcMainEvent, ...args: ExtractArgs<T>[E]) => void | Promise<void>
-  ): void {
-    this.listeners.push(channel)
+  ): () => void {
+    let listeners = this.listeners.get(channel)
+    if (!listeners) {
+      listeners = new Set()
+      this.listeners.set(channel, listeners)
+    }
+    listeners.add(listener)
     ipcMain.on(channel, listener as (e: Electron.IpcMainEvent, ...args: any[]) => void)
+    return () => {
+      ipcMain.removeListener(channel, listener)
+      listeners.delete(listener)
+      if (listeners.size === 0) {
+        this.listeners.delete(channel)
+      }
+    }
   }
 
   /**
    * Handle a renderer invoke request.
+   * @returns A function to remove the handler.
    */
   handle<E extends keyof ExtractHandler<T>>(
     channel: Extract<E, string>,
     listener: (
       e: Electron.IpcMainInvokeEvent,
       ...args: Parameters<ExtractHandler<T>[E]>
     ) => ReturnType<ExtractHandler<T>[E]> | Promise<ReturnType<ExtractHandler<T>[E]>>
-  ): void {
-    this.handlers.push(channel)
+  ): () => void {
+    this.handlers.add(channel)
     ipcMain.handle(channel, listener as (event: Electron.IpcMainInvokeEvent, ...args: any[]) => any)
+    return () => {
+      ipcMain.removeHandler(channel)
+      this.handlers.delete(channel)
+    }
   }
 
   /**
-   * Dispose all listeners and handlers.
+   * Dispose all registered listeners and handlers.
    */
   dispose(): void {
-    this.listeners.forEach(c => ipcMain.removeAllListeners(c))
-    this.listeners = []
-    this.handlers.forEach(c => ipcMain.removeHandler(c))
-    this.handlers = []
+    this.listeners.forEach((listeners, channel) => {
+      listeners.forEach(listener => ipcMain.removeListener(channel, listener))
+    })
+    this.listeners.clear()
+    this.handlers.forEach(channel => ipcMain.removeHandler(channel))
+    this.handlers.clear()
   }
 }
```

Comment:
- **Old behavior**: `dispose` removes _all_ listeners for each channel, even those which were not registered by the instance.
- **New behavior**: `dispose` removes listeners registered by this specific instance.

I think the new behavior is less surprising than the old behavior, but you could argue that this is a breaking change.